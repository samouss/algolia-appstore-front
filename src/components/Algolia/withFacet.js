import React, { Component } from 'react';
import { getDisplayName } from 'core/utils';
import { ContextTypes } from './Provider';

const withFacet = ({
  facet,
  maxFacetValues = 10,
  getFacetValuesOptions = {},
} = {}) => WrappedComponent => {
  if (!facet) {
    throw new Error('You must provide the facet parameter.');
  }

  class WithFacet extends Component {

    constructor(props) {
      super(props);

      this.state = {
        facetValues: [],
      };

      this.updateState = this.updateState.bind(this);
    }

    componentDidMount() {
      const { algoliaHelper } = this.context;

      algoliaHelper.on('result', this.updateState);
    }

    componentWillUnmount() {
      this.context.algoliaHelper.removeListener('result', this.updateState);
    }

    updateState(content) {
      const facetValues = content.getFacetValues(
        facet,
        getFacetValuesOptions,
      );

      const facetValuesUpToMaxNbValuesToDisplay = facetValues
        .filter(_ => _.count > 0)
        .slice(0, maxFacetValues);

      this.setState(() => ({
        facetValues: facetValuesUpToMaxNbValuesToDisplay,
      }));
    }

    render() {
      return (
        <WrappedComponent
          {...this.props}
          {...this.state}
        />
      );
    }

  }

  WithFacet.displayName = getDisplayName(
    WrappedComponent,
    'withFacet',
  );

  WithFacet.contextTypes = ContextTypes;

  return WithFacet;
};

export default withFacet;
