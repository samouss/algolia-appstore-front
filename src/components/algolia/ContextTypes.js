import PropTypes from 'prop-types';

export default {
  algoliaClient: PropTypes.object.isRequired,
  algoliaHelper: PropTypes.object.isRequired,
};
