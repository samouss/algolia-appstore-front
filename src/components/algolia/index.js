export { default as Provider } from './Provider';
export { default as withPaginateResults } from './withPaginateResults';
