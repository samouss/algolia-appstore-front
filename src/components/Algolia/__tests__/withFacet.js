import React from 'react';
import PropTypes from 'prop-types';
import { shallow } from 'enzyme';
import { createMockAlgoliaClient, createMockAlgoliaHelper } from 'test/algolia';
import withFacet from '../withFacet';

describe('algolia', () => {
  describe('withFacet', () => {
    const createContext = () => ({
      algoliaHelper: createMockAlgoliaHelper(),
      algoliaClient: createMockAlgoliaClient(),
    });

    const defaultParameters = {
      facet: 'category',
    };

    const Component = ({ children, className }) => (
      <div className={className}>{children}</div>
    );

    Component.propTypes = {
      className: PropTypes.string.isRequired,
      children: PropTypes.node.isRequired,
    };

    Component.defaultProps = {
      className: '',
    };

    it('expect to render', () => {
      const context = createContext();

      const parameters = {
        ...defaultParameters,
      };

      const props = {
        className: 'sample-class-name',
      };

      const ApplyComponent = withFacet(parameters)(Component);

      const component = shallow(
        <ApplyComponent
          {...props}
        >
          <div>Content</div>
        </ApplyComponent>,
        { context },
      );

      expect(component).toMatchSnapshot();
    });

    describe('componentDidMount', () => {
      it('expect to subscribe to result event', () => {
        const context = createContext();

        const parameters = {
          ...defaultParameters,
        };

        const ApplyComponent = withFacet(parameters)(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.instance().componentDidMount();

        expect(context.algoliaHelper.on).toHaveBeenCalledWith(
          'result',
          component.instance().updateState,
        );
      });
    });

    describe('componentWillUnmount', () => {
      it('expect to unsubscribe to result event', () => {
        const context = createContext();

        const parameters = {
          ...defaultParameters,
        };

        const ApplyComponent = withFacet(parameters)(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        component.instance().componentWillUnmount();

        expect(context.algoliaHelper.removeListener).toHaveBeenCalledWith(
          'result',
          component.instance().updateState,
        );
      });
    });

    describe('updateState', () => {
      it('expect to sort facet values by count desc', () => {
        const context = createContext();

        const parameters = {
          ...defaultParameters,
        };

        const ApplyComponent = withFacet(parameters)(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        const content = {
          getFacetValues: jest.fn(() => []),
        };

        component.instance().updateState(content);

        expect(content.getFacetValues).toHaveBeenCalledWith(
          parameters.facet,
          { sortBy: ['count:desc'] },
        );
      });

      it('expect to keep only facet values greater than 0', () => {
        const context = createContext();

        const parameters = {
          ...defaultParameters,
        };

        const ApplyComponent = withFacet(parameters)(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        const content = {
          getFacetValues: jest.fn(() => [
            { count: 0 },
            { count: 10 },
          ]),
        };

        const expectation = [
          { count: 10 },
        ];

        component.instance().updateState(content);

        expect(component.state().facetValues).toEqual(expectation);
      });

      it('expect to keep only given number of facet values', () => {
        const context = createContext();

        const parameters = {
          ...defaultParameters,
          maxFacetValues: 2,
        };

        const ApplyComponent = withFacet(parameters)(Component);

        const component = shallow(
          <ApplyComponent>
            <div>Content</div>
          </ApplyComponent>,
          { context },
        );

        const content = {
          getFacetValues: jest.fn(() => [
            { count: 30 },
            { count: 20 },
            { count: 10 },
          ]),
        };

        const expectation = [
          { count: 30 },
          { count: 20 },
        ];

        component.instance().updateState(content);

        expect(component.state().facetValues).toEqual(expectation);
      });
    });
  });
});
