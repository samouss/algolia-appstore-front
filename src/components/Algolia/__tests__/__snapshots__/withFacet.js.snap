// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`algolia withFacet connect expect to render 1`] = `
<withFacet(Component)
  className="sample-class-name"
  helper={
    Object {
      "addDisjunctiveFacetRefinement": [Function],
      "addNumericRefinement": [Function],
      "clearRefinements": [Function],
      "getIndex": [Function],
      "getState": [Function],
      "nextPage": [Function],
      "on": [Function],
      "removeDisjunctiveFacetRefinement": [Function],
      "removeListener": [Function],
      "removeNumericRefinement": [Function],
      "search": [Function],
      "setIndex": [Function],
      "setQuery": [Function],
      "setQueryParameter": [Function],
    }
  }
>
  <div>
    Content
  </div>
</withFacet(Component)>
`;

exports[`algolia withFacet expect to render 1`] = `
<Component
  className="sample-class-name"
  facetValues={Array []}
>
  <div>
    Content
  </div>
</Component>
`;
